File: src/code_indexing_mcp/progress.py
"""Live progress for an indexing run, in process and across processes.

An index can run for minutes, and the caller who has to wait it out is rarely
the process running it: the MCP server usually delegates to the per-user daemon,
and the CLI can be watching a run someone else started. So progress is both
handed to an in-process listener and published as a small JSON snapshot under
``<data>/progress/<project-id>.json`` that any process can read.

The snapshot is a status file, not a queue: every update replaces it, readers
only ever see the latest state, and a stale file left behind by a killed process
is ignored once it stops being refreshed.
"""

from __future__ import annotations

import contextlib
import os
import tempfile
import time
from collections.abc import Callable
from copy import deepcopy
from pathlib import Path

# IndexProgress lives in models.py (ProjectStatus embeds it); it is re-exported
# here because this module owns everything else about progress. Nothing in this
# module may be imported by models.py, or the import cycle this layout removed
# comes back.
from .models import IndexProgress as IndexProgress
from .models import IndexTrigger as IndexTrigger

# How long a snapshot stays trustworthy without a refresh. Comfortably above the
# publish interval, and above the pauses a single slow file can cause between
# updates, so a live run is never mistaken for a dead one.
STALE_AFTER_SECONDS = 60.0
PUBLISH_INTERVAL_SECONDS = 0.25


def progress_path(directory: Path, project_id: str) -> Path:
    return directory / f"{project_id}.json"


class ProgressPublisher:
    """Throttled writer for one run's snapshot file and in-process listener."""

    def __init__(
        self,
        project_id: str,
        *,
        run_id: str = "",
        trigger: IndexTrigger = "manual",
        directory: Path | None = None,
        listener: Callable[[IndexProgress], None] | None = None,
        interval_seconds: float = PUBLISH_INTERVAL_SECONDS,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        self.directory = directory
        self.listener = listener
        self.interval_seconds = interval_seconds
        self._clock = clock
        self._last_published: float | None = None
        self.state = IndexProgress(
            project_id=project_id,
            run_id=run_id,
            trigger=trigger,
            started_at=time.time(),
            phase_started_at=self._clock(),
        )

    @property
    def enabled(self) -> bool:
        return self.directory is not None or self.listener is not None

    def update(self, *, force: bool = False, **fields: object) -> None:
        """Merge *fields* into the snapshot, publishing at most once per interval.

        Publishing is never allowed to break an index: a full disk or a racing
        reader costs an update, not the run.
        """

        if not self.enabled:
            return
        phase = fields.get("phase")
        if phase is not None and phase != self.state.phase:
            # Anchor the new phase strictly after the previous one: clocks with
            # coarse granularity (some Windows runners read milliseconds) can
            # return the same tick twice, which would make the anchor not
            # advance and phase durations collapse to zero.
            fields["phase_started_at"] = max(self._clock(), self.state.phase_started_at + 1e-6)
        # Deep-copy the merged fields: model_copy never copies the update
        # mapping's values, and a published snapshot must not share a nested
        # value (e.g. the skipped_by_reason dict) with a caller that keeps
        # mutating it. Every retained snapshot stays a true point-in-time
        # picture.
        self.state = self.state.model_copy(update=deepcopy(fields))
        now = self._clock()
        if (
            not force
            and self._last_published is not None
            and now - self._last_published < self.interval_seconds
        ):
            return
        self._last_published = now
        self.state.updated_at = time.time()
        if self.listener is not None:
            self.listener(self.state)
        if self.directory is not None:
            with contextlib.suppress(OSError):
                self._write(self.state)

    def clear(self) -> None:
        """Remove the snapshot once the run is over."""

        if self.directory is None:
            return
        with contextlib.suppress(OSError):
            progress_path(self.directory, self.state.project_id).unlink(missing_ok=True)

    def _write(self, state: IndexProgress) -> None:
        assert self.directory is not None
        self.directory.mkdir(parents=True, exist_ok=True)
        payload = state.model_dump_json()
        # Written whole and renamed into place so a reader polling the file
        # never parses a half-written snapshot.
        descriptor, temporary = tempfile.mkstemp(dir=self.directory, suffix=".tmp")
        try:
            with os.fdopen(descriptor, "w") as handle:
                handle.write(payload)
            os.replace(temporary, progress_path(self.directory, state.project_id))
        except BaseException:
            Path(temporary).unlink(missing_ok=True)
            raise


def read_progress(
    directory: Path, project_id: str, *, stale_after_seconds: float = STALE_AFTER_SECONDS
) -> IndexProgress | None:
    """Return the latest snapshot for *project_id*, or None when there is none.

    A snapshot older than *stale_after_seconds* is treated as absent: it was
    left behind by a process that died rather than by one still working.
    """

    path = progress_path(directory, project_id)
    try:
        payload = path.read_text()
    except (OSError, ValueError):
        return None
    try:
        progress = IndexProgress.model_validate_json(payload)
    except ValueError:
        return None
    if time.time() - progress.updated_at > stale_after_seconds:
        return None
    return progress

File: ts/packages/server/src/progress.ts
/**
 * Live progress for an indexing run, in process and across processes.
 *
 * An index can run for minutes, and the caller who has to wait it out is rarely
 * the process running it: the MCP server usually delegates to the per-user
 * daemon, and the CLI can be watching a run someone else started. So progress is
 * both handed to an in-process listener and published as a small JSON snapshot
 * under `<data>/progress/<project-id>.json` that any process can read.
 *
 * The snapshot is a status file, not a queue: every update replaces it, readers
 * only ever see the latest state, and a stale file left behind by a killed
 * process is ignored once it stops being refreshed.
 */

import fs from "node:fs";
import path from "node:path";
// IndexProgress lives in models.ts (ProjectStatus embeds it); it is re-exported
// here because this module owns everything else about progress. Nothing in this
// module may be imported by models.ts, or the import cycle this layout removed
// comes back.
import { IndexProgress, type IndexTrigger } from "./models.ts";

export { IndexProgress, type IndexTrigger };

/**
 * How long a snapshot stays trustworthy without a refresh. Comfortably above the
 * publish interval, and above the pauses a single slow file can cause between
 * updates, so a live run is never mistaken for a dead one.
 */
export const STALE_AFTER_SECONDS = 60;
export const PUBLISH_INTERVAL_SECONDS = 0.25;

export function progressPath(directory: string, projectId: string): string {
  return path.join(directory, `${projectId}.json`);
}

/** Seconds since the epoch, the unit the snapshot's timestamps are written in. */
function wallClockSeconds(): number {
  return Date.now() / 1000;
}

/** `time.monotonic()`: seconds from an arbitrary origin that never goes back. */
function monotonicSeconds(): number {
  return performance.now() / 1000;
}

export interface ProgressPublisherOptions {
  readonly runId?: string;
  readonly trigger?: IndexTrigger;
  readonly directory?: string | undefined;
  readonly listener?: ((progress: IndexProgress) => void) | undefined;
  readonly intervalSeconds?: number;
  readonly clock?: () => number;
}

/** Throttled writer for one run's snapshot file and in-process listener. */
export class ProgressPublisher {
  readonly directory: string | undefined;
  readonly listener: ((progress: IndexProgress) => void) | undefined;
  readonly intervalSeconds: number;
  state: IndexProgress;

  readonly #clock: () => number;
  #lastPublished: number | null = null;

  constructor(projectId: string, options: ProgressPublisherOptions = {}) {
    this.directory = options.directory;
    this.listener = options.listener;
    this.intervalSeconds = options.intervalSeconds ?? PUBLISH_INTERVAL_SECONDS;
    this.#clock = options.clock ?? monotonicSeconds;
    this.state = IndexProgress.parse({
      project_id: projectId,
      run_id: options.runId ?? "",
      trigger: options.trigger ?? "manual",
      started_at: wallClockSeconds(),
      phase_started_at: this.#clock(),
    });
  }

  get enabled(): boolean {
    return this.directory !== undefined || this.listener !== undefined;
  }

  /**
   * Merge *fields* into the snapshot, publishing at most once per interval.
   *
   * Publishing is never allowed to break an index: a full disk or a racing
   * reader costs an update, not the run.
   */
  update(fields: Partial<IndexProgress>, { force = false }: { force?: boolean } = {}): void {
    if (!this.enabled) return;
    const merged: Partial<IndexProgress> = { ...fields };
    if (merged.phase !== undefined && merged.phase !== this.state.phase) {
      // Anchor the new phase strictly after the previous one: clocks with coarse
      // granularity (some Windows runners read milliseconds) can return the same
      // tick twice, which would make the anchor not advance and phase durations
      // collapse to zero.
      merged.phase_started_at = Math.max(this.#clock(), this.state.phase_started_at + 1e-6);
    }
    // Deep-copy the merged fields, and replace the state object rather than
    // mutating it: a published snapshot must not share a nested value (the
    // skipped_by_reason record, say) with a caller that keeps mutating it, and a
    // listener that retains what it was handed must keep a true point-in-time
    // picture.
    this.state = { ...this.state, ...structuredClone(merged) };
    const now = this.#clock();
    if (
      !force &&
      this.#lastPublished !== null &&
      now - this.#lastPublished < this.intervalSeconds
    ) {
      return;
    }
    this.#lastPublished = now;
    this.state = { ...this.state, updated_at: wallClockSeconds() };
    this.listener?.(this.state);
    if (this.directory !== undefined) {
      try {
        writeSnapshot(this.directory, this.state);
      } catch {
        // An unwritable snapshot is not the run's problem.
      }
    }
  }

  /** Remove the snapshot once the run is over. */
  clear(): void {
    if (this.directory === undefined) return;
    try {
      fs.rmSync(progressPath(this.directory, this.state.project_id), { force: true });
    } catch {
      // Same contract as update(): losing the file is never worth an exception.
    }
  }
}

/**
 * Write the snapshot whole and rename it into place, so a reader polling the
 * file never parses a half-written one.
 */
function writeSnapshot(directory: string, state: IndexProgress): void {
  fs.mkdirSync(directory, { recursive: true });
  const target = progressPath(directory, state.project_id);
  const temporary = path.join(directory, `.${state.project_id}.${process.pid}.${nextSeq()}.tmp`);
  try {
    fs.writeFileSync(temporary, JSON.stringify(state), { flag: "wx" });
    fs.renameSync(temporary, target);
  } catch (error) {
    fs.rmSync(temporary, { force: true });
    throw error;
  }
}

let sequence = 0;
function nextSeq(): number {
  sequence += 1;
  return sequence;
}

/**
 * Return the latest snapshot for *projectId*, or null when there is none.
 *
 * A snapshot older than *staleAfterSeconds* is treated as absent: it was left
 * behind by a process that died rather than by one still working.
 */
export function readProgress(
  directory: string,
  projectId: string,
  { staleAfterSeconds = STALE_AFTER_SECONDS }: { staleAfterSeconds?: number } = {},
): IndexProgress | null {
  let progress: IndexProgress;
  try {
    progress = IndexProgress.parse(
      JSON.parse(fs.readFileSync(progressPath(directory, projectId), "utf8")),
    );
  } catch {
    return null;
  }
  if (wallClockSeconds() - progress.updated_at > staleAfterSeconds) return null;
  return progress;
}
